### ISSUE TITLE
Allow debugging test

### ISSUE THREAD

**Issue by PasVegan** [FalsePattern/ZigBrains#82]
It would be nice to be able to debug tests.

**Comment by FalsePattern** [comment]
Debugging single-file tests is already supported.

**Comment by FalsePattern** [comment]
If you mean debugging tests launched using `zig build test`, it is unfortunately too complicated for me to implement, but I'm open for contributions.

**Comment by PasVegan** [comment]
> If you mean debugging tests launched using `zig build test`, it is unfortunately too complicated for me to implement for the foreseeable future, but I'm open for contributions.

Yes that's what I meant, if I have some time I will try to help 👍🏻

**Comment by g41797** [comment]
You can debug single file test only if the test in question does not contain dependencies.
It's restriction of 'zig test' - it does not use build system.

In order to debud tests, I 
- [install test exe](https://github.com/g41797/syslog/blob/main/build.zig#L62)
- use [zig build install](https://github.com/g41797/syslog/blob/main/.run/testall.run.xml)

**Comment by Bastczuak** [comment]
> If you mean debugging tests launched using `zig build test`, it is unfortunately too complicated for me to implement for the foreseeable future, but I'm open for contributions.

Can you please explain why it is too complicated. I am asking to know where to start too help you and tackle this feature.

**Comment by FalsePattern** [comment]
The main issue comes from the fact that inspecting task outputs for finding test binaries requires a custom build runner, at that would be a large maintenance burden.

In the latest version of zigbrains, you can do it semi-manually though, by creating an install step for the test binary, and then setting up the task for it.

Example for the `zig init` template repo for zig 0.14.0:

Add this to the end of `build.zig`'s build function:
```zig
    const test_artifact = b.addInstallArtifact(
        exe_unit_tests,
        .{ .dest_dir = .{ .override = .{ .custom = "tests" } } },
    );
    const install_test_step = b.step("install_test", "Create test binaries for debugging");
    install_test_step.dependOn(&test_artifact.step);
```

This generates the following file when run:
![Image](https://github.com/user-attachments/assets/80efe51b-dda1-4c0d-9383-99fd1a1c3538)

Next, create a new `Zig build` intellij run configuration, and set it up like so:
![Image](https://github.com/user-attachments/assets/8d0330c1-623f-43f4-bcd9-13b3b40d3e2c)

And when you debug this configuration, the test executable generated by the `install_test` task gets sent to gdb correctly:
![Image](https://github.com/user-attachments/assets/6c5297c6-da7f-42cc-9026-31dfc5395d82)

**Comment by FalsePattern** [comment]
Only works for single-executable tests though, so if you have multiple `b.addTest(...)` tests, you need to set up an install step + intellij run config for each

**Comment by aohanhongzhi** [comment]
![Image](https://github.com/user-attachments/assets/019a5842-627b-4312-a105-23bc03c70025)

**Comment by FalsePattern** [comment]
> ![Image](https://github.com/user-attachments/assets/019a5842-627b-4312-a105-23bc03c70025)

Yes, scroll up, there's a guide on how to configure your build.zig to support debugging tests

**Comment by FalsePattern** [comment]
@aohanhongzhi scroll up, there's a tutorial on how to modify your build.zig to support debugging tests

**Comment by Vivraan** [comment]
Hey @FalsePattern, hope it's good in 2026!

I tried the steps from your tutorial with these modifications (CLion 2025.3.1, Windows 11):

## `build.zig`
```zig
    // ...

    const mod_tests = b.addTest(.{
        .root_module = mod,
    });

    const install_mod_tests = b.addInstallArtifact(mod_tests, .{ .dest_dir = .{ .override = .{ .custom = "test_root" } } });

    const mod_tests_step = b.step("test_root", "Create test binaries for debugging 'root'");
    mod_tests_step.dependOn(&install_mod_tests.step);

    const exe_tests = b.addTest(.{
        .root_module = exe.root_module,
    });

    const install_exe_tests = b.addInstallArtifact(exe_tests, .{ .dest_dir = .{ .override = .{ .custom = "test_exe" } } });

    const exe_tests_step = b.step("test_exe", "Create test binaries for debugging the main module");
    exe_tests_step.dependOn(&install_exe_tests.step);
}
// eof
```

## Run configuration

I created one for the root module and one for the executable, as [suggested](https://github.com/FalsePattern/ZigBrains/issues/82#issuecomment-2758859230):

<img width="990" height="857" alt="Image" src="https://github.com/user-attachments/assets/e5c848c4-09ef-49ac-93d1-a272cf249651" />

For `Test Exe` I just replace mentions of `_root` with `_exe` in all the fields and the path.

## Debug Run Output

```
C:\Users\shiva\AppData\Local\JetBrains\CLion2025.3\zigbrains\msvc\VsDebugConsole.exe \\.\pipe\Microsoft-VisualStudio-Debug-Console-{0E2E11E2-0DC1-4DBA-976A-E570F2D169C5} Local\{0E2E11E2-0DC1-4DBA-976A-E570F2D169C5} /AutoExit
All 0 tests passed.

Process finished with exit code 0
```

This wasn't the case when I was simply running `zig build test` as a run artefact. In fact, no debug points in my code get triggered at all, and stuff that's clearly meant to print to `stderr` via `std.debug.print` don't appear.

Is there anything I'm missing here that may have changed since the last IDE or toolchain updates?

I understand that this isn't a first-class or even planned feature, but if this thread can act as documentation for it, I'm happy to assist in elaborating this use case.

**Comment by g41797** [comment]
I am talking about CLion 2025.3.1 under **_Fedora 43 Kde Plasma_**

[CLion debug configuration](https://github.com/g41797/tofu/blob/main/.run/testall.run.xml#L7) - install step/test (exe)

Zig debugged under GDB

additional "fine tuning" _after update of CLion_ - adding 
```zig
        .use_llvm = true,
        .use_lld = true,
```
everywhere in [build.zig](https://github.com/g41797/tofu/blob/main/build.zig) in order to force usage of LLVM as backend

**Comment by Vivraan** [comment]
Adding to this again.

Running a single-file test magically runs all tests, even for files not on the run config path. I'm assuming it pulls tests from dependencies.

<img width="1000" height="857" alt="Image" src="https://github.com/user-attachments/assets/51de6f1c-97e1-4617-8635-b5f4584691fd" />

Config XML:
```xml
<component name="ProjectRunConfigurationManager">
  <configuration default="false" name="test VM.zig" type="ZIGBRAINS_TEST" factoryName="ZIGBRAINS_TEST">
    <ZigBrainsOption name="workingDirectory" value="O:\CraftingInterpreters\ziglox" />
    <envs />
    <ZigBrainsOption name="filePath" value="O:\CraftingInterpreters\ziglox\src\VM.zig" />
    <ZigBrainsNestedOption name="optimization">
      <ZigBrainsOption name="level" value="Debug" />
      <ZigBrainsOption name="forced" value="false" />
    </ZigBrainsNestedOption>
    <ZigBrainsOption name="compilerArgs" value="" />
    <method v="2" />
  </configuration>
</component>
```

Clarifying again that this is MSVC, should I switch to LLVM for this?

**Comment by Vivraan** [comment]
I think I have lost the plot. Here are links to my current setup [`build test` ❌ ; `test xyz.zig` ✅]:
https://gitlab.com/Vivraan/chaka/-/tree/66af3e361a414a8424b59a82137bc39d4bf73987/ziglox/.idea/runConfigurations
https://gitlab.com/Vivraan/chaka/-/blob/66af3e361a414a8424b59a82137bc39d4bf73987/ziglox/build.zig
These point to a specific commit, so there may be future changes that I'll have to reconfirm if this thread is updated again.

However, I can't step through code in either case. I can only trigger the debugger in single-file tests.

**Comment by Vivraan** [comment]
Just updating for completion: I resolved my case by just adding tests. [More info](https://ziggit.dev/t/how-to-debug-zig-with-lldb-on-windows/13862?u=vivraan).
